h1st/core/graph.py

```diff
@@ -0,0 +1,79 @@
+import collections
+
+from h1st.core import NodeInfo, GraphInfo
+from h1st.core.execution import ExecutionManager
+from h1st.core.node import Node
+
+
+class Graph:
+    def __init__(self) -> None:
+        self.__graph = {}  # dictionary of Node -> NodeInfo
+        self.__root_nodes: list[Node] = []
+        self.__last_updated_node = None
+
+    def add_node(self, node: Node):
+        if not isinstance(node, Node):
+            raise TypeError("must be Node, not {}".format(type(node).__name__))
+        if node not in self.__graph.keys():
+            self.__graph[node] = NodeInfo()
+
+    def add_edge(self, from_node: Node, to_node: Node):
+        if not isinstance(from_node, Node):
+            raise TypeError("must be Node, not {}".format(type(from_node).__name__))
+        if not isinstance(to_node, Node):
+            raise TypeError("must be Node, not {}".format(type(to_node).__name__))
+        is_root_node = to_node in self.__root_nodes
+        if is_root_node and len(self.__root_nodes) == 1:
+            raise RuntimeError("There must be one or more root nodes")
+        if from_node not in self.__graph.keys():
+            self.add_node(from_node)
+        if to_node not in self.__graph.keys():
+            self.add_node(to_node)
+        if from_node in self.__graph[to_node].next_nodes:
+            raise RuntimeError("Cannot add edge as it results in a closed loop")
+        if to_node not in self.__graph[from_node].next_nodes and \
```

Comment:
this one can we use `elif to_node not in self.__graph[from_node].next_nodes` to reduce length of if condition?